src/flash.rs

```diff
@@ -2,6 +2,275 @@
+
+        // Write address bits
+        // NOTE(unsafe) This sets the sector address in the Address Register
+        unsafe { self.flash.ar.ar().write(|w| w.far().bits(FLASH_START + start_offset) ); }
+
+        // Start Operation
+        self.flash.cr.cr().modify(|_, w| w.strt().set_bit() );
+
+        // Wait for operation to finish
+        while self.flash.sr.sr().read().bsy().bit_is_set() { }
+
+        // Check for errors
+        let sr = self.flash.sr.sr().read();
+
+        // Remove Page Erase Operation bit
+        self.flash.cr.cr().modify(|_, w| w.per().clear_bit() );
+
+        // Re-lock flash
+        self.lock()?;
+
+        if sr.wrprterr().bit_is_set() {
+            self.flash.sr.sr().modify(|_, w| w.wrprterr().clear_bit() );
+            Err(Error::EraseError)
+        } else {
+            if self.verify {
+                let start = start_offset & !(self.sector_sz.kbytes() as u32 -1);
+                let end = start + self.sector_sz.kbytes() as u32 -1;
+                for idx in start..end {
+                    let write_address = (FLASH_START + idx as u32) as *const u16;
+                    let verify: u16 = unsafe { core::ptr::read_volatile(write_address) };
+                    if verify != 0xFFFF {
+                        return Err(Error::VerifyError);
+                    }
+                }
+            }
+
+            Ok(())
+        }
+    }
+
+    /// Erase the Flash Sectors from FLASH_START + start_offset to length
+    pub fn erase(&mut self, start_offset: u32, length: usize) -> Result<()> {
+        self.valid_length(start_offset, length)?;
+
+        // Erase every sector touched by start_offset + length
+        for offset in (start_offset..start_offset + length as u32).step_by(self.sector_sz.kbytes() as usize) {
+            self.page_erase(offset)?;
+        }
+
+        // Report Success
+        Ok(())
+    }
+
+    /// Retrieve an slice of data from FLASH_START +offset
+    pub fn read(&self, offset: u32, length: usize) -> Result<FlashReaderGuard> {
+        self.valid_address(offset)?;
+
+        if offset + length as u32 > self.flash_sz.kbytes() as u32 {
+            return Err(Error::LengthTooLong);
+        }
+
+        let address = (FLASH_START + offset) as *const _;
+
+        Ok(FlashReaderGuard {
+            _flash: self,
+
+            // NOTE(unsafe) read with no side effects safe-guarded by FlashReaderGuard.
+            data: unsafe { core::slice::from_raw_parts::<'static, u8>(address, length) }
+        })
+    }
+
+    /// Write data to FLASH_START +offset
+    pub fn write(&mut self, offset: u32, data: &[u8]) -> Result<()> {
+        self.valid_length(offset, data.len() )?;
+
+        // Unlock Flash
+        self.unlock()?;
+
+        for idx in (0..data.len() ).step_by(2) {
+            self.valid_address(offset + idx as u32)?;
+
+            let hword: u16 =
+                  (data[idx]   as u16)
+                | (data[idx+1] as u16) << 8;
```

Comment:
Weird formatting. Though I'm not sure what the best way to format it is

Reply:
it's better to use `copy_from_slice` and `u16::from_be/le_bytes` for such manipulations